File: js/src/telemetry.ts
/**
 * 사용량 텔레메트리 — 어느 증권사가 얼마나 쓰이는지 시간 버킷으로 합산해 `hermetix-service` 로 보낸다.
 * 계약(필드·전송 규칙·보내지 않는 것)은 `docs/telemetry.md` 가 정본이다 (Kotlin UsageTelemetry 와 동일 의미).
 *
 * - 매매 경로와 분리: 카운터는 메모리, 전송은 unref 타이머. 실패는 조용히 버리고 큐를 쌓지 않는다
 * - 개인정보·매매 내용 없음: 브로커·환경·호출 종류·건수·에러 분류·응답 시간 분포·스트림 건수·SDK 버전·설치 ID 뿐
 * - 기본 배포본은 항상 켜져 있다 (설정 없음). 테스트는 `transport` 와 `flushNow()` 로 전송을 가로챈다
 */
import { createHmac, randomUUID } from "node:crypto";
import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { createRequire } from "node:module";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import type { BrokerClient } from "./broker.js";
import {
  AuthError, BrokerApiError, InsufficientFundsError, InvalidOrderError, MarketClosedError, OrderNotFoundError, RateLimitError,
} from "./errors.js";
import type { StreamChannel, TradingEnvironment } from "./models.js";

export const TELEMETRY_ENDPOINT = "https://service-api-prod.hermetix.dev/v1/usage";
export const TELEMETRY_SCHEMA = 1;
export const SDK_LANGUAGE = "js";
/** 요청 서명 키 (docs/telemetry.md "요청 서명") — 공개 SDK 라 비밀이 아니며 스팸·스캐너를 거르는 문턱이다 */
export const SIGNING_KEY_ID = "v1";
export const SIGNING_KEY = "d97f20cb942540462ea83648ee30a9786bd658b3dc813f74ef011845da258503";
/** 전송 대상 — 테스트에서만 `__setEndpointForTests` 로 바꾼다 (사용자 설정 없음) */
let endpoint: string = TELEMETRY_ENDPOINT;
export function __setEndpointForTests(url: string | null): void { endpoint = url ?? TELEMETRY_ENDPOINT; }

/** `hex(HMAC-SHA256(key, timestamp + "\n" + body))` — 계약의 요청 서명 */
export function signTelemetry(body: string, timestampSeconds: number): string {
  return createHmac("sha256", SIGNING_KEY).update(`${timestampSeconds}\n${body}`, "utf8").digest("hex");
}
const FLUSH_INTERVAL_MS = 60_000;
const LATENCY_SAMPLES = 256;
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export type ErrorClass =
  | "rate_limit" | "auth" | "market_closed" | "insufficient_funds" | "invalid_order" | "order_not_found" | "network" | "other";

function readSdkVersion(): string {
  try {
    const require = createRequire(import.meta.url);
    // dist/src/telemetry.js → ../../package.json, src/telemetry.ts → ../package.json
    for (const candidate of ["../../package.json", "../package.json"]) {
      try {
        const pkg = require(candidate) as { name?: string; version?: string };
        if (pkg.name === "hermetix" && pkg.version) return pkg.version;
      } catch { /* 다음 후보 */ }
    }
  } catch { /* 무시 */ }
  return "unknown";
}

function loadOrCreateInstallationId(): string {
  try {
    const path = join(homedir(), ".hermetix", "installation-id");
    if (existsSync(path)) {
      const id = readFileSync(path, "utf-8").trim();
      if (UUID_RE.test(id)) return id;
    }
    const id = randomUUID();
    try { mkdirSync(dirname(path), { recursive: true }); writeFileSync(path, id); } catch { /* 못 쓰면 임시 ID */ }
    return id;
  } catch {
    return randomUUID();
  }
}

/** op 당 최근 256개 표본만 보관 — p50/p95 는 전송 시점에 계산 */
class LatencyReservoir {
  private readonly samples = new Array<number>(LATENCY_SAMPLES);
  private count = 0;
  add(ms: number): void { this.samples[this.count % LATENCY_SAMPLES] = ms; this.count++; }
  summary(): { count: number; p50: number; p95: number } {
    const n = Math.min(this.count, LATENCY_SAMPLES);
    if (n === 0) return { count: 0, p50: 0, p95: 0 };
    const sorted = this.samples.slice(0, n).sort((a, b) => a - b);
    const pct = (p: number) => sorted[Math.min(n - 1, Math.max(0, Math.floor((n - 1) * p)))];
    return { count: this.count, p50: pct(0.5), p95: pct(0.95) };
  }
}

class OpStats { ok = 0; readonly errors = new Map<ErrorClass, number>(); readonly latency = new LatencyReservoir(); }
class StreamStats { subscriptions = 0; messages = 0; }
class BucketStats {
  readonly ops = new Map<string, OpStats>();
  readonly streams = new Map<string, StreamStats>();
  reconnects = 0;
  toJson(hour: string, broker: string, environment: string) {
    return {
      hour, broker, environment,
      ops: [...this.ops.entries()].sort(([a], [b]) => a.localeCompare(b)).map(([op, s]) => ({
        op, ok: s.ok, errors: Object.fromEntries(s.errors), latencyMs: s.latency.summary(),
      })),
      streams: [...this.streams.entries()].sort(([a], [b]) => a.localeCompare(b)).map(([channel, s]) => ({
        channel, subscriptions: s.subscriptions, messages: s.messages,
      })),
      reconnects: this.reconnects,
    };
  }
}

/** 예외를 계약의 에러 분류로 (docs/telemetry.md) */
export function classifyError(err: unknown): ErrorClass {
  if (err instanceof RateLimitError) return "rate_limit";
  if (err instanceof AuthError) return "auth";
  if (err instanceof MarketClosedError) return "market_closed";
  if (err instanceof InsufficientFundsError) return "insufficient_funds";
  if (err instanceof InvalidOrderError) return "invalid_order";
  if (err instanceof OrderNotFoundError) return "order_not_found";
  if (err instanceof BrokerApiError) return "other";
  if (err && typeof err === "object") {
    const e = err as { name?: string; code?: string; cause?: unknown; message?: string };
    if (e.name === "AbortError" || e.name === "TimeoutError") return "network";
    if (typeof e.code === "string" && /^(ECONN|ENOTFOUND|ETIMEDOUT|EAI_AGAIN|EPIPE|UND_ERR)/.test(e.code)) return "network";
    if (err instanceof TypeError && /fetch failed|network|socket|ECONN/i.test(e.message ?? "")) return "network";
    if (e.cause !== undefined && e.cause !== err) return classifyError(e.cause);
  }
  return "other";
}

const buckets = new Map<string, { hour: string; broker: string; environment: string; stats: BucketStats }>();
let sdkVersionCache: string | null = null;
let installationIdCache: string | null = null;
let timer: NodeJS.Timeout | null = null;

function bucketFor(broker: string, environment: TradingEnvironment): BucketStats {
  const hour = new Date(Math.floor(Date.now() / 3_600_000) * 3_600_000).toISOString().replace(".000Z", "Z");
  const key = `${hour}|${broker}|${environment}`;
  let b = buckets.get(key);
  if (!b) { b = { hour, broker, environment, stats: new BucketStats() }; buckets.set(key, b); }
  return b.stats;
}

async function postDefault(body: string): Promise<void> {
  const controller = new AbortController();
  const t = setTimeout(() => controller.abort(), 3_000);
  t.unref?.();
  try {
    const timestamp = Math.floor(Date.now() / 1000);
    await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "User-Agent": `hermetix-${SDK_LANGUAGE}/${UsageTelemetry.sdkVersion}`,
        "X-Hermetix-Key-Id": SIGNING_KEY_ID,
        "X-Hermetix-Timestamp": String(timestamp),
        "X-Hermetix-Signature": signTelemetry(body, timestamp),
      },
      body,
      signal: controller.signal,
    });
  } finally {
    clearTimeout(t);
  }
}

export const UsageTelemetry = {
  get sdkVersion(): string { return (sdkVersionCache ??= readSdkVersion()); },
  get installationId(): string { return (installationIdCache ??= loadOrCreateInstallationId()); },

  /** 전송 함수 — 기본은 fetch POST(3초 타임아웃, 응답 버림). 테스트에서 교체한다 */
  transport: postDefault as (body: string) => void | Promise<void>,

  /** 어댑터가 자기 브로커·환경으로 만들어 두고 쓰는 핸들 */
  forBroker(brokerId: string, environment: TradingEnvironment): BrokerUsage { return new BrokerUsage(brokerId, environment); },

  record(broker: string, environment: TradingEnvironment, op: string, ms: number, failure: unknown | null): void {
    const stats = bucketFor(broker, environment);
    let s = stats.ops.get(op);
    if (!s) { s = new OpStats(); stats.ops.set(op, s); }
    if (failure === null || failure === undefined) s.ok++;
    else { const c = classifyError(failure); s.errors.set(c, (s.errors.get(c) ?? 0) + 1); }
    s.latency.add(Math.max(0, Math.round(ms)));
  },

  recordStream(broker: string, environment: TradingEnvironment, channel: StreamChannel, subscriptions: number, messages: number): void {
    const stats = bucketFor(broker, environment);
    let s = stats.streams.get(channel);
    if (!s) { s = new StreamStats(); stats.streams.set(channel, s); }
    s.subscriptions += subscriptions;
    s.messages += messages;
  },

  recordReconnect(broker: string, environment: TradingEnvironment): void { bucketFor(broker, environment).reconnects++; },

  /** 지금까지 쌓인 버킷을 페이로드 JSON 으로 만들고 비운다. 비어 있으면 null */
  drain(now: Date = new Date()): string | null {
    if (buckets.size === 0) return null;
    const list = [...buckets.values()];
    buckets.clear();
    return JSON.stringify({
      schema: TELEMETRY_SCHEMA,
      installationId: this.installationId,
      sdk: { language: SDK_LANGUAGE, version: this.sdkVersion },
      sentAt: now.toISOString().replace(/\.\d{3}Z$/, "Z"),
      buckets: list.map((b) => b.stats.toJson(b.hour, b.broker, b.environment)),
    });
  },

  /** 즉시 전송 시도 (타이머·종료·테스트용). 비어 있으면 아무것도 안 하고, 전송 실패는 삼킨다 — 절대 reject 하지 않는다 */
  async flushNow(): Promise<void> {
    const body = this.drain();
    if (body === null) return;
    try { await this.transport(body); } catch { /* 실패한 페이로드는 다시 보내지 않는다 */ }
  },
};

function startScheduler(): void {
  if (timer) return;
  timer = setInterval(() => { void UsageTelemetry.flushNow(); }, FLUSH_INTERVAL_MS);
  timer.unref?.();
  process.once("beforeExit", () => { void UsageTelemetry.flushNow(); });
}

/**
 * 브로커 어댑터 하나가 쥐는 계측 핸들. 공개 메서드를 [measure] 로 감싸면 건수·에러 분류·응답 시간이 쌓인다.
 * 동기·비동기 함수 모두 지원하며 기록은 finally 에서 하므로 어떤 경로로 나가도 세어진다.
 */
export class BrokerUsage {
  constructor(readonly brokerId: string, readonly environment: TradingEnvironment) { startScheduler(); }

  measure<T>(op: string, fn: () => T): T {
    const start = performance.now();
    let result: T;
    try {
      result = fn();
    } catch (e) {
      UsageTelemetry.record(this.brokerId, this.environment, op, performance.now() - start, e);
      throw e;
    }
    if (result instanceof Promise) {
      return result.then(
        (v) => { UsageTelemetry.record(this.brokerId, this.environment, op, performance.now() - start, null); return v; },
        (e) => { UsageTelemetry.record(this.brokerId, this.environment, op, performance.now() - start, e); throw e; },
      ) as unknown as T;
    }
    UsageTelemetry.record(this.brokerId, this.environment, op, performance.now() - start, null);
    return result;
  }

  streamSubscribed(channel: StreamChannel, count = 1): void { if (count !== 0) UsageTelemetry.recordStream(this.brokerId, this.environment, channel, count, 0); }
  streamMessage(channel: StreamChannel, count = 1): void { UsageTelemetry.recordStream(this.brokerId, this.environment, channel, 0, count); }
  reconnected(): void { UsageTelemetry.recordReconnect(this.brokerId, this.environment); }
}

const PUBLIC_OPS: [keyof BrokerClient, string][] = [
  ["getQuotes", "quotes"], ["getCandles", "candles"], ["getCalendar", "calendar"], ["getAccount", "account"],
  ["getHoldings", "holdings"], ["getBuyingPower", "buying_power"], ["createOrder", "create_order"],
  ["getOrders", "get_orders"], ["getOrder", "get_order"], ["cancelOrder", "cancel_order"], ["getFills", "fills"],
];
const INSTRUMENTED = new WeakSet<object>();

/**
 * 어댑터 인스턴스의 공개 메서드 11개를 계측으로 감싼다 (정확히 한 번). 어댑터 생성자 끝에서 호출해
 * 사용자가 클라이언트를 직접 만들어도 항상 계측된다. 돌려주는 핸들은 토큰 발급(`auth`)·스트림 카운트에 쓴다.
 */
export function instrumentBroker(client: BrokerClient): BrokerUsage {
  const usage = UsageTelemetry.forBroker(client.capabilities.brokerId, client.environment ?? "PAPER");
  if (INSTRUMENTED.has(client)) return usage;
  INSTRUMENTED.add(client);
  const target = client as unknown as Record<string, (...args: unknown[]) => unknown>;
  for (const [method, op] of PUBLIC_OPS) {
    const original = target[method];
    if (typeof original !== "function") continue;
    target[method] = function (this: unknown, ...args: unknown[]) { return usage.measure(op, () => original.apply(this, args)); };
  }
  return usage;
}

File: go/telemetry.go
package hermetix

// 사용량 텔레메트리 — 어느 증권사가 얼마나 쓰이는지 시간 버킷으로 합산해 hermetix-service 로 보낸다 (Kotlin UsageTelemetry 대응).
// 계약(필드·전송 규칙·보내지 않는 것)은 docs/telemetry.md 가 정본이다.
//
//   - 매매 경로와 분리: 카운터는 메모리, 전송은 백그라운드 고루틴. 실패는 조용히 버리고 큐를 쌓지 않는다
//   - 개인정보·매매 내용 없음: 브로커·환경·호출 종류·건수·에러 분류·응답 시간 분포·스트림 건수·SDK 버전·설치 ID 뿐
//   - 기본 배포본은 항상 켜져 있다 (설정 없음). 테스트는 TelemetryTransport 와 FlushTelemetry 로 전송을 가로챈다
//   - Go 에는 종료 훅이 없으므로 프로세스가 끝나기 전에 FlushTelemetry() 를 부르면 남은 버킷을 한 번 더 보낸다 (StrategyEngine.Stop 이 부른다)

import (
	"bytes"
	"context"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"sync"
	"time"
)

const (
	// TelemetryEndpoint - 수신 엔드포인트. 언어별로 상수 한 곳에만 있으며 사용자 설정은 없다
	TelemetryEndpoint = "https://service-api-prod.hermetix.dev/v1/usage"
	// TelemetrySchema - 계약 버전
	TelemetrySchema = 1
	telemetrySDK    = "go"

	// TelemetrySigningKeyID / TelemetrySigningKey - 요청 서명 (docs/telemetry.md "요청 서명").
	// 공개 SDK 라 비밀이 아니며, 아무 curl 이나 스캐너가 보내지 못하게 하는 문턱이다. 키 회전 시 ID 를 올린다
	TelemetrySigningKeyID = "v1"
	TelemetrySigningKey   = "d97f20cb942540462ea83648ee30a9786bd658b3dc813f74ef011845da258503"

	telemetryFlushInterval = 60 * time.Second
	telemetryLatencySample = 256
)

// TelemetryTransport - 전송 함수. 기본은 HTTP POST. 테스트에서 교체한다.
var TelemetryTransport = telemetryPost

type telemetryBucketKey struct {
	hour        time.Time
	broker      string
	environment string
}

type telemetryOpStats struct {
	ok      int64
	errors  map[string]int64
	samples [telemetryLatencySample]int64
	count   int64
}

type telemetryStreamStats struct {
	subscriptions int64
	messages      int64
}

type telemetryBucket struct {
	ops        map[string]*telemetryOpStats
	streams    map[string]*telemetryStreamStats
	reconnects int64
}

type telemetryState struct {
	mu             sync.Mutex
	buckets        map[telemetryBucketKey]*telemetryBucket
	startOnce      sync.Once
	installationID string
	idOnce         sync.Once
}

var telemetry = &telemetryState{buckets: map[telemetryBucketKey]*telemetryBucket{}}

// BrokerUsage - 브로커 어댑터 하나가 쥐는 계측 핸들.
//
//	func (c *X) GetQuotes(...) (_ []Quote, err error) {
//	    defer c.usage().Measure("quotes")(&err)
//	    ...
//	}
type BrokerUsage struct {
	BrokerID    string
	Environment TradingEnvironment
}

// Measure - 호출 시작 시각을 잡고, 돌려준 함수를 defer 로 부르면 건수·에러 분류·응답 시간이 기록된다.
func (u BrokerUsage) Measure(op string) func(err *error) {
	start := time.Now()
	return func(err *error) {
		var failure error
		if err != nil {
			failure = *err
		}
		RecordUsage(u.BrokerID, u.Environment, op, time.Since(start), failure)
	}
}

// StreamSubscribed - 구독(등록) 수를 더한다.
func (u BrokerUsage) StreamSubscribed(channel StreamChannel, count int) {
	if count > 0 {
		RecordStreamUsage(u.BrokerID, u.Environment, channel, int64(count), 0)
	}
}

// StreamMessage - 리스너에 전달한 틱/이벤트 수를 더한다.
func (u BrokerUsage) StreamMessage(channel StreamChannel, count int) {
	if count > 0 {
		RecordStreamUsage(u.BrokerID, u.Environment, channel, 0, int64(count))
	}
}

// Reconnected - 웹소켓 재접속 1회.
func (u BrokerUsage) Reconnected() { RecordReconnect(u.BrokerID, u.Environment) }

// RecordUsage - 어댑터 계측용 — 보통 BrokerUsage.Measure 를 통해 호출된다.
func RecordUsage(brokerID string, environment TradingEnvironment, op string, elapsed time.Duration, failure error) {
	telemetry.ensureStarted()
	telemetry.mu.Lock()
	defer telemetry.mu.Unlock()
	b := telemetry.bucketLocked(brokerID, environment)
	s := b.ops[op]
	if s == nil {
		s = &telemetryOpStats{errors: map[string]int64{}}
		b.ops[op] = s
	}
	if failure == nil {
		s.ok++
	} else {
		s.errors[ClassifyTelemetryError(failure)]++
	}
	s.samples[int(s.count%telemetryLatencySample)] = elapsed.Milliseconds()
	s.count++
}

// RecordStreamUsage - 스트림 구독·메시지 수.
func RecordStreamUsage(brokerID string, environment TradingEnvironment, channel StreamChannel, subscriptions, messages int64) {
	telemetry.ensureStarted()
	telemetry.mu.Lock()
	defer telemetry.mu.Unlock()
	b := telemetry.bucketLocked(brokerID, environment)
	s := b.streams[string(channel)]
	if s == nil {
		s = &telemetryStreamStats{}
		b.streams[string(channel)] = s
	}
	s.subscriptions += subscriptions
	s.messages += messages
}

// RecordReconnect - 웹소켓 재접속 횟수.
func RecordReconnect(brokerID string, environment TradingEnvironment) {
	telemetry.ensureStarted()
	telemetry.mu.Lock()
	defer telemetry.mu.Unlock()
	telemetry.bucketLocked(brokerID, environment).reconnects++
}

// ClassifyTelemetryError - 예외를 계약의 에러 분류로 (docs/telemetry.md).
func ClassifyTelemetryError(err error) string {
	if err == nil {
		return "other"
	}
	var (
		rate  *RateLimitError
		auth  *AuthError
		mc    *MarketClosedError
		funds *InsufficientFundsError
		inv   *InvalidOrderError
		nf    *OrderNotFoundError
		api   *BrokerAPIError
		nerr  net.Error
		uerr  *url.Error
	)
	switch {
	case errors.As(err, &rate):
		return "rate_limit"
	case errors.As(err, &auth):
		return "auth"
	case errors.As(err, &mc):
		return "market_closed"
	case errors.As(err, &funds):
		return "insufficient_funds"
	case errors.As(err, &inv):
		return "invalid_order"
	case errors.As(err, &nf):
		return "order_not_found"
	case errors.As(err, &api):
		return "other"
	case errors.As(err, &nerr), errors.As(err, &uerr),
		errors.Is(err, context.DeadlineExceeded), errors.Is(err, io.ErrUnexpectedEOF), errors.Is(err, io.EOF):
		return "network"
	}
	return "other"
}

// DrainTelemetry - 지금까지 쌓인 버킷을 페이로드 JSON 으로 만들고 비운다. 비어 있으면 nil.
func DrainTelemetry(now time.Time) []byte {
	telemetry.mu.Lock()
	buckets := telemetry.buckets
	telemetry.buckets = map[telemetryBucketKey]*telemetryBucket{}
	telemetry.mu.Unlock()
	if len(buckets) == 0 {
		return nil
	}
	keys := make([]telemetryBucketKey, 0, len(buckets))
	for k := range buckets {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		if !keys[i].hour.Equal(keys[j].hour) {
			return keys[i].hour.Before(keys[j].hour)
		}
		if keys[i].broker != keys[j].broker {
			return keys[i].broker < keys[j].broker
		}
		return keys[i].environment < keys[j].environment
	})
	out := make([]map[string]any, 0, len(keys))
	for _, k := range keys {
		out = append(out, buckets[k].toJSON(k))
	}
	payload := map[string]any{
		"schema":         TelemetrySchema,
		"installationId": TelemetryInstallationID(),
		"sdk":            map[string]string{"language": telemetrySDK, "version": Version},
		"sentAt":         now.UTC().Format(time.RFC3339),
		"buckets":        out,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return nil
	}
	return body
}

// FlushTelemetry - 즉시 전송 시도. 비어 있으면 아무것도 안 하고, 전송 실패는 삼킨다 (재전송 없음).
// 프로세스 종료 전에 한 번 부르면 남은 버킷이 나간다 — StrategyEngine.Stop 이 부른다.
func FlushTelemetry() {
	body := DrainTelemetry(time.Now())
	if body == nil {
		return
	}
	defer func() {
		if r := recover(); r != nil {
			log.Printf("DEBUG hermetix telemetry: send skipped - %v", r)
		}
	}()
	if err := TelemetryTransport(body); err != nil {
		log.Printf("DEBUG hermetix telemetry: send skipped - %v", err)
	}
}

// TelemetryInstallationID - ~/.hermetix/installation-id 의 UUID. 못 읽고 못 쓰면 프로세스 수명 동안만 유지되는 임시 ID.
func TelemetryInstallationID() string {
	telemetry.idOnce.Do(func() { telemetry.installationID = loadOrCreateInstallationID() })
	return telemetry.installationID
}

// ------------------------------------------------------------------ internals

func (t *telemetryState) ensureStarted() {
	t.startOnce.Do(func() {
		go func() {
			ticker := time.NewTicker(telemetryFlushInterval)
			defer ticker.Stop()
			for range ticker.C {
				FlushTelemetry()
			}
		}()
	})
}

func (t *telemetryState) bucketLocked(brokerID string, environment TradingEnvironment) *telemetryBucket {
	key := telemetryBucketKey{hour: time.Now().UTC().Truncate(time.Hour), broker: brokerID, environment: string(environment)}
	b := t.buckets[key]
	if b == nil {
		b = &telemetryBucket{ops: map[string]*telemetryOpStats{}, streams: map[string]*telemetryStreamStats{}}
		t.buckets[key] = b
	}
	return b
}

func (b *telemetryBucket) toJSON(key telemetryBucketKey) map[string]any {
	opNames := make([]string, 0, len(b.ops))
	for op := range b.ops {
		opNames = append(opNames, op)
	}
	sort.Strings(opNames)
	ops := make([]map[string]any, 0, len(opNames))
	for _, op := range opNames {
		s := b.ops[op]
		errs := map[string]int64{}
		for k, v := range s.errors {
			errs[k] = v
		}
		ops = append(ops, map[string]any{"op": op, "ok": s.ok, "errors": errs, "latencyMs": s.latencySummary()})
	}
	chNames := make([]string, 0, len(b.streams))
	for ch := range b.streams {
		chNames = append(chNames, ch)
	}
	sort.Strings(chNames)
	streams := make([]map[string]any, 0, len(chNames))
	for _, ch := range chNames {
		s := b.streams[ch]
		streams = append(streams, map[string]any{"channel": ch, "subscriptions": s.subscriptions, "messages": s.messages})
	}
	return map[string]any{
		"hour": key.hour.Format(time.RFC3339), "broker": key.broker, "environment": key.environment,
		"ops": ops, "streams": streams, "reconnects": b.reconnects,
	}
}

// latencySummary - op 당 최근 256개 표본으로 p50/p95 를 계산한다 (count 는 전체 건수).
func (s *telemetryOpStats) latencySummary() map[string]int64 {
	n := int(s.count)
	if n > telemetryLatencySample {
		n = telemetryLatencySample
	}
	if n == 0 {
		return map[string]int64{"count": 0, "p50": 0, "p95": 0}
	}
	sorted := make([]int64, n)
	copy(sorted, s.samples[:n])
	sort.Slice(sorted, func(i, j int) bool { return sorted[i] < sorted[j] })
	pct := func(p float64) int64 {
		i := int(float64(n-1) * p)
		if i < 0 {
			i = 0
		}
		if i > n-1 {
			i = n - 1
		}
		return sorted[i]
	}
	return map[string]int64{"count": s.count, "p50": pct(0.50), "p95": pct(0.95)}
}

var telemetryHTTP = &http.Client{
	Timeout:   3 * time.Second,
	Transport: &http.Transport{DialContext: (&net.Dialer{Timeout: 2 * time.Second}).DialContext},
}

// telemetryEndpoint - 실제 전송 주소. 상수 TelemetryEndpoint 와 같으며 테스트에서만 바꾼다
var telemetryEndpoint = TelemetryEndpoint

// SignTelemetry - 계약의 요청 서명: hex(HMAC-SHA256(key, timestamp + "\n" + body)), 소문자 hex 64자
func SignTelemetry(body []byte, timestampSeconds int64) string {
	mac := hmac.New(sha256.New, []byte(TelemetrySigningKey))
	mac.Write([]byte(fmt.Sprintf("%d\n", timestampSeconds)))
	mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

func telemetryPost(body []byte) error {
	req, err := http.NewRequest("POST", telemetryEndpoint, bytes.NewReader(body))
	if err != nil {
		return err
	}
	timestamp := time.Now().Unix()
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("User-Agent", "hermetix-"+telemetrySDK+"/"+Version)
	req.Header.Set("X-Hermetix-Key-Id", TelemetrySigningKeyID)
	req.Header.Set("X-Hermetix-Timestamp", fmt.Sprintf("%d", timestamp))
	req.Header.Set("X-Hermetix-Signature", SignTelemetry(body, timestamp))
	resp, err := telemetryHTTP.Do(req)
	if err != nil {
		return err
	}
	_, _ = io.Copy(io.Discard, resp.Body)
	_ = resp.Body.Close()
	return nil
}

var uuidPattern = regexp.MustCompile(`^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$`)

func newUUIDv4() string {
	var b [16]byte
	if _, err := rand.Read(b[:]); err != nil {
		now := time.Now().UnixNano()
		for i := range b {
			b[i] = byte(now >> (uint(i%8) * 8))
		}
	}
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

func loadOrCreateInstallationID() string {
	home, err := os.UserHomeDir()
	if err != nil || home == "" {
		return newUUIDv4()
	}
	path := filepath.Join(home, ".hermetix", "installation-id")
	if data, err := os.ReadFile(path); err == nil {
		if id := strings.TrimSpace(string(data)); uuidPattern.MatchString(strings.ToLower(id)) {
			return strings.ToLower(id)
		}
	}
	id := newUUIDv4()
	if err := os.MkdirAll(filepath.Dir(path), 0o700); err == nil {
		_ = os.WriteFile(path, []byte(id), 0o600)
	}
	return id
}
